drivers/saul/init_devs/auto_init_saul_pwm.c

```diff
@@ -0,0 +1,179 @@
+ * 1kHz, and the indicated channels are exposed via SAUL.
+ *
+ * @}
+ */
+
+#include "log.h"
+#include "saul_reg.h"
+#include "saul/periph.h"
+#include "pwm_params.h"
+#include "periph/pwm.h"
+
+#define MAX(a, b) (((a) > (b)) ? (a) : (b))
+
+#if !defined(SAUL_PWM_FREQ)
+#define SAUL_PWM_FREQ SAUL_PWM_FREQ_DEFAULT
+#endif
+
+/**
+ * @brief   Define the number of configured dimmers
+ */
+#ifndef SAUL_PWM_NO_DIMMER
+#define SAUL_PWM_DIMMER_NUMOF ARRAY_SIZE(saul_pwm_dimmer_params)
+#else
+#define SAUL_PWM_DIMMER_NUMOF 0
+static const saul_pwm_dimmer_params_t saul_pwm_dimmer_params[1];
+#endif
+
+/**
+ * @brief   Define the number of configured RGB LEDs
+ */
+#ifndef SAUL_PWM_NO_RGB
+#define SAUL_PWM_RGB_NUMOF ARRAY_SIZE(saul_pwm_rgb_params)
+#else
+#define SAUL_PWM_RGB_NUMOF 0
+static const saul_pwm_rgb_params_t saul_pwm_rgb_params[1];
+#endif
+
+/**
+ * @brief   Memory for the registry RGB LED entries
+ */
+/* The static variable will be unused in the 0 case and thus not emitted. */
+static saul_reg_t saul_reg_entries_rgb[MAX(SAUL_PWM_RGB_NUMOF, 1)];
+
+/**
+ * @brief   Memory for the registry dimmer entries
+ */
+/* The static variable will be unused in the 0 case and thus not emitted. */
+static saul_reg_t saul_reg_entries_dimmer[MAX(SAUL_PWM_DIMMER_NUMOF, 1)];
+
+/**
+ * @brief   Reference to the driver for single-channel dimmers
+ */
+extern saul_driver_t dimmer_saul_driver;
+
+/**
+ * @brief   Reference to the driver for RGB LEDs
+ */
+extern saul_driver_t rgb_saul_driver;
+
+/**
+ * Configure a PWM driver for LED output (1kHz, 8bit)
+ */
+static int configure(pwm_t dev)
+{
+    LOG_DEBUG("[auto_init_saul] initializing PWM %u for LED operation,", dev);
+    uint32_t freq = pwm_init(dev, PWM_LEFT, SAUL_PWM_FREQ, saul_pwm_resolution);
+    LOG_DEBUG(" actual frequency %lu,\n", freq);
+    return freq != 0 ? 0 : -ENOTSUP;
+}
+
+/**
+ * Configure the PWM driver at the given index (inside saul_pwm_dimmer_params,
+ * overflowing into saul_pwm_rgb_params) unless that device came up previously,
+ * in which case the function returns without any action.
+ * */
+static int configure_on_first_use(pwm_t dev, int index)
+{
+    for (int i = 0; i < (int)SAUL_PWM_DIMMER_NUMOF; ++i) {
+        pwm_t currentdev = saul_pwm_dimmer_params[i].channel.dev;
+        if (currentdev == dev) {
+            if (i == index) {
+                return configure(dev);
+            }
+            return 0;
+        }
+    }
+    for (int i = 0; i < (int)SAUL_PWM_RGB_NUMOF; ++i) {
+        for (int j = 0; j < 3; ++j) {
+            int index = SAUL_PWM_DIMMER_NUMOF + i * 3 + j;
+            pwm_t currentdev = saul_pwm_rgb_params[i].channels[j].dev;
+            if (currentdev == dev) {
+                if (i == index) {
+                    return configure(dev);
+                }
+                return 0;
+            }
+        }
+    }
+    return -ENOENT;
+}
+
+void auto_init_pwm(void)
+{
+    for (int i = 0; i < (int)SAUL_PWM_DIMMER_NUMOF; i++) {
```

Comment:
```suggestion
    for (usigned i = 0; i < SAUL_PWM_DIMMER_NUMOF; i++) {
```

This also matches the `%u` format specifier below in `LOG_ERROR()`.